File: src/devtwin/adapters/rust.py
"""Rust ecosystem adapter."""

from __future__ import annotations

import re
from pathlib import Path

from devtwin.adapters.base import EcosystemAdapter, extract_version
from devtwin.core.models import DependencyInfo, HealthIssue, Presence, RuntimeInfo, Severity
from devtwin.core.runner import run_command, which

TOOLCHAIN_CHANNEL_RE = re.compile(r'channel\s*=\s*"([^"]+)"')


class RustAdapter(EcosystemAdapter):
    ecosystem = "rust"

    def detect(self, root: Path) -> bool:
        return (root / "Cargo.toml").exists()

    def _required_version(self, root: Path) -> tuple[str | None, str | None]:
        for filename in ("rust-toolchain.toml", "rust-toolchain"):
            f = root / filename
            if not f.exists():
                continue
            try:
                text = f.read_text(encoding="utf-8", errors="ignore")
            except OSError:
                continue
            match = TOOLCHAIN_CHANNEL_RE.search(text)
            if match:
                return match.group(1), filename
            stripped = text.strip()
            if stripped and filename == "rust-toolchain":
                return stripped, filename
        return None, None

    def inspect_runtime(self, root: Path) -> list[RuntimeInfo]:
        required, source = self._required_version(root)
        rustc_path = which("rustc")
        if rustc_path is None:
            return [
                RuntimeInfo(
                    name="rustc",
                    presence=Presence.NOT_INSTALLED,
                    required_version=required,
                    source=source,
                )
            ]
        result = run_command([rustc_path, "--version"], timeout=5)
        installed = extract_version(result.stdout) if result.available else None
        return [
            RuntimeInfo(
                name="rustc",
                presence=Presence.INSTALLED if installed else Presence.UNKNOWN,
                installed_version=installed,
                required_version=required,
                path=rustc_path,
                source=source,
            )
        ]

    def inspect_dependencies(self, root: Path) -> DependencyInfo | None:
        if not (root / "Cargo.toml").exists():
            return None
        lockfile = root / "Cargo.lock"
        return DependencyInfo(
            ecosystem="rust",
            manager="cargo",
            lockfile="Cargo.lock" if lockfile.exists() else None,
            lockfile_present=lockfile.exists(),
            manifest_present=True,
        )

    def inspect_build(self, root: Path) -> list[RuntimeInfo]:
        cargo_path = which("cargo")
        if cargo_path is None:
            return [RuntimeInfo(name="cargo", presence=Presence.NOT_INSTALLED)]
        result = run_command([cargo_path, "--version"], timeout=5)
        installed = extract_version(result.stdout) if result.available else None
        return [
            RuntimeInfo(
                name="cargo",
                presence=Presence.INSTALLED,
                installed_version=installed,
                path=cargo_path,
            )
        ]

    def inspect_tests(self, root: Path) -> list[str]:
        return ["cargo test"] if (root / "Cargo.toml").exists() else []

    def inspect_build_commands(self, root: Path) -> list[str]:
        return ["cargo build"] if (root / "Cargo.toml").exists() else []

    def health_checks(self, root: Path, runtimes: list[RuntimeInfo]) -> list[HealthIssue]:
        issues = []
        for rt in runtimes:
            if rt.name == "rustc" and rt.presence == Presence.NOT_INSTALLED:
                issues.append(
                    HealthIssue(
                        severity=Severity.CRITICAL,
                        code="rust.not_installed",
                        title="Rust toolchain not found",
                        message="No `rustc` executable was found on PATH.",
                        evidence=["which rustc -> not found"],
                        recommendation="Install Rust via https://rustup.rs.",
                    )
                )
        return issues

File: mcphub/devtwin/src/adapters/rust.ts
/**
 * Rust ecosystem adapter.
 */

import { join } from 'node:path';

import {
  Presence,
  Severity,
  type DependencyInfo,
  type HealthIssue,
  type RuntimeInfo,
} from '../core/models.js';
import { runCommand, which } from '../core/runner.js';
import { pathExists, readTextFile } from '../system/filesystem.js';
import { EcosystemAdapter, extractVersion } from './base.js';

const TOOLCHAIN_CHANNEL_RE = /channel\s*=\s*"([^"]+)"/;

export class RustAdapter extends EcosystemAdapter {
  readonly ecosystem = 'rust';

  detect(root: string): boolean {
    return pathExists(join(root, 'Cargo.toml'));
  }

  /**
   * Targeted line/regex read of the toolchain file -- no TOML parser needed:
   * we only ever want the `channel` key, and the bare `rust-toolchain` form
   * is a plain version string on its own.
   */
  private requiredVersion(root: string): [string | null, string | null] {
    for (const filename of ['rust-toolchain.toml', 'rust-toolchain']) {
      const f = join(root, filename);
      if (!pathExists(f)) {
        continue;
      }
      const text = readTextFile(f);
      if (text === null) {
        continue;
      }
      const match = TOOLCHAIN_CHANNEL_RE.exec(text);
      if (match) {
        return [match[1]!, filename];
      }
      const stripped = text.trim();
      if (stripped && filename === 'rust-toolchain') {
        return [stripped, filename];
      }
    }
    return [null, null];
  }

  override async inspectRuntime(root: string): Promise<RuntimeInfo[]> {
    const [required, source] = this.requiredVersion(root);
    const rustcPath = which('rustc');
    if (rustcPath === null) {
      return [
        {
          name: 'rustc',
          presence: Presence.NOT_INSTALLED,
          installed_version: null,
          required_version: required,
          path: null,
          source,
        },
      ];
    }
    const result = await runCommand([rustcPath, '--version'], { timeout: 5 });
    const installed = result.available ? extractVersion(result.stdout) : null;
    return [
      {
        name: 'rustc',
        presence: installed ? Presence.INSTALLED : Presence.UNKNOWN,
        installed_version: installed,
        required_version: required,
        path: rustcPath,
        source,
      },
    ];
  }

  override async inspectDependencies(root: string): Promise<DependencyInfo | null> {
    if (!pathExists(join(root, 'Cargo.toml'))) {
      return null;
    }
    const lockfilePresent = pathExists(join(root, 'Cargo.lock'));
    return {
      ecosystem: 'rust',
      manager: 'cargo',
      lockfile: lockfilePresent ? 'Cargo.lock' : null,
      lockfile_present: lockfilePresent,
      manifest_present: true,
      installed: Presence.UNKNOWN,
      dependency_count: null,
      dev_dependency_count: null,
      notes: [],
    };
  }

  override async inspectBuild(_root: string): Promise<RuntimeInfo[]> {
    const cargoPath = which('cargo');
    if (cargoPath === null) {
      return [
        {
          name: 'cargo',
          presence: Presence.NOT_INSTALLED,
          installed_version: null,
          required_version: null,
          path: null,
          source: null,
        },
      ];
    }
    const result = await runCommand([cargoPath, '--version'], { timeout: 5 });
    const installed = result.available ? extractVersion(result.stdout) : null;
    return [
      {
        name: 'cargo',
        presence: Presence.INSTALLED,
        installed_version: installed,
        required_version: null,
        path: cargoPath,
        source: null,
      },
    ];
  }

  override async inspectTests(root: string): Promise<string[]> {
    return pathExists(join(root, 'Cargo.toml')) ? ['cargo test'] : [];
  }

  override async inspectBuildCommands(root: string): Promise<string[]> {
    return pathExists(join(root, 'Cargo.toml')) ? ['cargo build'] : [];
  }

  override async healthChecks(_root: string, runtimes: RuntimeInfo[]): Promise<HealthIssue[]> {
    const issues: HealthIssue[] = [];
    for (const rt of runtimes) {
      if (rt.name === 'rustc' && rt.presence === Presence.NOT_INSTALLED) {
        issues.push({
          severity: Severity.CRITICAL,
          code: 'rust.not_installed',
          title: 'Rust toolchain not found',
          message: 'No `rustc` executable was found on PATH.',
          evidence: ['which rustc -> not found'],
          recommendation: 'Install Rust via https://rustup.rs.',
          confidence: null,
        });
      }
    }
    return issues;
  }
}
